### PR TITLE
Add SOCKS proxy support to the API

### PR DESCRIPTION
**Pull request by carterkozak**
==COMMIT_MSG==
Add SOCKS proxy support to the API
==COMMIT_MSG==

## Possible downsides?
more proxies, socks is very hard to unit test

### CODE REVIEW

File: service-config/src/test/java/com/palantir/conjure/java/api/config/service/ProxyConfigurationTests.java [hunk 1]
```diff
@@ -117,6 +121,17 @@ public void credentialsWithNonHttp() {
...
 
+    @Test
+    public void credentialsWithSocksProxy() {
+        assertThatThrownBy(() -> ProxyConfiguration.builder()
+                        .type(ProxyConfiguration.Type.SOCKS)
+                        .credentials(BasicCredentials.of("foo", "bar"))
+                        .hostAndPort("localhost:1234")
+                        .build())
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("credentials only valid for HTTP proxies");
```
Comment (CRogers) [inline service-config/src/test/java/com/palantir/conjure/java/api/config/service/ProxyConfigurationTests.java:132]: Hmm, can't socks proxies have credentials? Or is this a matter of just restricting the API for now?

Comment (CRogers) [reply service-config/src/test/java/com/palantir/conjure/java/api/config/service/ProxyConfigurationTests.java:132]: Seemingly they can but it seems a little scary with passwords sent in cleartext: https://www.rfc-editor.org/rfc/rfc1929

**Comment by CRogers** [comment]
:+!:

**Comment by CRogers** [comment]
:+1:

File: service-config/src/test/java/com/palantir/conjure/java/api/config/service/ProxyConfigurationTests.java (hunk 1 above)
Comment (carterkozak) [reply service-config/src/test/java/com/palantir/conjure/java/api/config/service/ProxyConfigurationTests.java:132]: This is _very_ difficult to test, so I opted to limit the scope to the minimal feature set we need.

**Pull request merged**